### PR TITLE
Fix for issue 2200: support new use-case for BLOCKCLONE backup method…

### PR DESCRIPTION
**Pull request by petroniusniger**
… by attempting to unmount device before backup/restore.

* Type:**Enhancement**

* Impact: **Normal**

* Reference to related issue (URL): https://github.com/rear/rear/issues/2200

* How was this pull request tested?
    * test VM installed with Leap 15.0 (grub2-efi, LVM, filesystems are ext4 and xfs)
    * created new, LUKS-encrypted logical volume with xfs mounted on /products
    * stored an extra decryption passphrase into slot #1 of LUKS header (main key in slot #0) 
    * created a "multiple backup" configuration where the base system is backed up the usual way (ISO + NETFS to NFS share), ignoring /products, while the block device that hosts /products is then backed up (mkbackuponly) using the BLOCKCLONE method. See configuration files below:
[site.conf.txt](https://github.com/rear/rear/files/3466879/site.conf.txt)
[base_system.conf.txt](https://github.com/rear/rear/files/3466880/base_system.conf.txt)
[products_backup.conf.txt](https://github.com/rear/rear/files/3466881/products_backup.conf.txt)
    * the new option `BLOCKCLONE_TRY_UNMOUNT` is set to "yes"
    * during /products backup phase (mkbackuponly), the filesystem is now successfully unmounted prior to issuing the `dd` command
    * at the end of the backup phase, that filesystem is also successfully re-mounted
    * the test VM is then shut down and its virtual HDD is deleted and recreated empty (same size, in this case)
    * during base system recover phase, the encrypted filesystem is correctly recreated (the user is prompted to enter a new passphrase -- which will later be overwritten). No data is restored on that filesystem as we ignored it through `BACKUP_PROG_EXCLUDE`
    * during /products restore phase (restoreonly) the target device (`/dev/vg00/lvol4`) is successfully unmounted prior to issuing the `dd` command, which overwrites the recreated encrypted filesystem with the original LUKS keys and all the filesystem data
    * upon reboot of the VM, all filesystems are clean, /products can be decrypted using any of the 2 original passphrases and can be successfully mounted

* Brief description of the changes in this pull request:
    * new configuration variable `BLOCKCLONE_TRY_UNMOUNT` in `default.conf`, including short documentation
    * `500_start_clone.sh` now use existing global function `umount_mountpoint()` to try and unmount the source device if `BLOCKCLONE_TRY_UNMOUNT` is set to "yes". It obtains the mount point through a new global function called `get_mountpoint()`
    * if the unmounting fails, `500_start_clone.sh` proceeds as before, issuing a warning message that the backup has been taken while the source device was mounted
    * if the device was initially mounted AND it could be unmounted, then `500_start_clone.sh` attempts to re-mounted it before exiting
    * `400_restore_clone.sh` now implements the same checks and unmount attempt as `500_start_clone.sh`. It does NOT try to remount the target device at the end, but issues a warning message instead
    * bug fix in global function `umount_mountpoint()`: return value now visible by calling script
    * cosmetics in global function `is_device_mounted()`: local variables now defined as `local`
    * new global function `get_mountpoint()`
    * extensive documentation of the new `BLOCKCLONE` use-case in the User Guide, chapter 12
    * fixed typo in User Guide, chapter 11, "multiple backups"

### CODE REVIEW

File: usr/share/rear/lib/global-functions.sh [hunk 1]
```diff
@@ -622,7 +622,8 @@ umount_mountpoint() {
     Log "Unmounting '$mountpoint'"
     umount $v $mountpoint >&2
     if [[ $? -eq 0 ]] ; then
-        return 0
+        echo 0
+	return
```
Comment (jsmeix) [inline usr/share/rear/lib/global-functions.sh]: From plain looking at the code here I think in general
all 'is_something' testing functions
should not provide the result as a string on stdout
but as usual via the exit code return value so that a testing function
could be used as usual by calling
```
is_something && ...
is_something || ...
```

**Comment by jsmeix** [comment]
@gozora 
could you have a look here?

FYI:
I am not in the office for some weeks so that I cannot do much for ReaR,
in particular I cannot try out something.

**Comment by gozora** [comment]
I'll take a look as time permits (most probably during this weekend) ...

V.

File: usr/share/rear/lib/global-functions.sh [hunk 2]
```diff
@@ -632,11 +633,13 @@ umount_mountpoint() {
     Log "Forced unmount of '$mountpoint'"
     umount $v -f $mountpoint >&2
     if [[ $? -eq 0 ]] ; then
-        return 0
+        echo 0
```
[the remark this answers is not in the document]
Comment (petroniusniger) [reply usr/share/rear/lib/global-functions.sh]: Hello @gozora 
You're right, of course. My apologies for missing this. I looked for calls to `umount_mountpoint()` throughout the scripts, but missed the fact that it could also be called by other functions.
I'll revert this change and adapt my own code accordingly.

File: usr/share/rear/backup/BLOCKCLONE/default/500_start_clone.sh [hunk 3]
```diff
@@ -32,3 +42,10 @@ case "$(basename ${BLOCKCLONE_PROG})" in
 esac
 
 StopIfError "Failed to create archive with $BLOCKCLONE_SOURCE_DEV"
+
+# If $BLOCKCLONE_SOURCE_DEV was initially mounted AND successfully unmounted, 
+# try to remount it before leaving
+if [ "$umount_res" = "0" ]; then
+    LogPrint "Trying to remount $mp"
+    mount $v $mp
```
[the remark this answers is not in the document]
Comment (petroniusniger) [reply usr/share/rear/backup/BLOCKCLONE/default/500_start_clone.sh]: Yes, I was aware of the fact that it would only work for mountpoints defined in `/etc/fstab`. I took the view that anything not present there would be transient anyway and that silently failing to remount it was acceptable behaviour -- if far from ideal.
But the approach you suggest is better by far, so I'll try to implement that ;-)

**Comment by petroniusniger** [comment]
Updated code to fulfil both suggestions by @gozora. 

Short comment about fixes in `is_device_mounted()` function:
- local variables now defined as local (was already the case in my first commit)
- fixed an 'inverse logic' bug in the test for empty parameter: not mounted or not found should return 0, not 1
- added missing early exit to the test for empty parameter -- passing an empty argument to the function might have resulted in an error similar to the following:
```
./fn_is_mounted: line 10: [: [SWAP]: binary operator expected
```

All of this does nothing to address the (valid) remark of @jsmeix about the behaviour of `is_something()` functions, but as I call `is_device_mounted()` from the code I contributed, I thought it fair game to fix these bugs at this time.

Regarding testing:
- I've tested extensively all three functions I modified or added by taking them out of ReaR and by feeding them different values to trigger a positive or negative result, including empty strings
- I've again performed a full "2-steps backups + 2-steps recover" scenario on a VM with a LUKS-encrypted filesystem. Everything behaved as expected.

Branch `issue-2200` has been re-based against `trunk/HEAD` before commit/push.

File: usr/share/rear/lib/global-functions.sh [hunk 4]
```diff
@@ -655,12 +655,12 @@ function change_default
 # Check if block device is mounted
 # lsblk can discover mounted device even if mounted as link, this makes it
 # more suitable for job then e.g. grep from /proc/mounts
-function is_device_mounted
+function is_device_mounted()
 {
-   disk=$1
-   [ -z "$disk" ] && echo 1
+   local disk=$1
+   [ -z "$disk" ] && echo 0 && return
```
Comment (gozora) [inline usr/share/rear/lib/global-functions.sh]: Thanks for fixing this!

File: usr/share/rear/restore/BLOCKCLONE/default/400_restore_clone.sh [hunk 5]
```diff
@@ -3,23 +3,56 @@
...
+fi
+
+local umount_res="-1"
+local mp=""
+if is_true "$BLOCKCLONE_TRY_UNMOUNT" && [ "$is_mounted" = "1" ]; then
+    mp=$(get_mountpoint $BLOCKCLONE_SOURCE_DEV)
+    # try unmount
+    if [ ! -z "$mp" ]; then
+        umount_mountpoint $mp
+	umount_res=$?
```
Comment (gozora) [inline usr/share/rear/restore/BLOCKCLONE/default/400_restore_clone.sh]: Please use correct indentation.

**Comment by gozora** [comment]
@petroniusniger thanks for updated PR. 
Please fix those couple indentation problems, and I'll merge this PR.

Thanks!

V.

File: usr/share/rear/backup/BLOCKCLONE/default/500_start_clone.sh [hunk 6]
```diff
@@ -9,12 +9,25 @@ using $BLOCKCLONE_PROG"
...
 
+local umount_res="-1"
+if is_true "$BLOCKCLONE_TRY_UNMOUNT" && [ "$is_mounted" = "1" ]; then
+    local mp=$(get_mountpoint $BLOCKCLONE_SOURCE_DEV)
+    # try unmount
+    if [ ! -z "$mp" ]; then
+        # save mount parameters for later
+	local mount_cmd=$(build_remount_cmd $mp)
+	umount_mountpoint $mp
+	umount_res=$?
```
[the remark this answers is not in the document]
Comment (petroniusniger) [reply usr/share/rear/backup/BLOCKCLONE/default/500_start_clone.sh]: Done.

File: usr/share/rear/restore/BLOCKCLONE/default/400_restore_clone.sh (hunk 5 above)
Comment (petroniusniger) [reply usr/share/rear/restore/BLOCKCLONE/default/400_restore_clone.sh]: Done.

**Comment by gozora** [comment]
@petroniusniger many thanks for contributing this new functionality and respective documentation to ReaR!

V.
